Feat: 장바구니 기능 구현

junga970 opened the pull request:
- #5

### 변경사항
<!-- 이 PR에서 어떤점들이 변경되었는지 기술해주세요. 가급적이면 as-is, to-be를 활용해서 작성해주세요.  -->
### **📌 AS-IS**
**1. 응답, 예외 처리**
- 기본적으로 ApiResponse으로 응답합니다.
- 예외 처리는 ResponseEntity<ErrorResponse>로 응답합니다.

<br>

**2. 장바구니에 대한 모든 기능은 로그인된 유저만 사용 가능**
- @AuthenticationPrincipal을 통해 로그인한 사용자 정보를 받습니다.

<br>

**3. 장바구니에 도서 담기 기능 추가**
- book id, 수량 정보를 받습니다.
- 이미 장바구니에 담긴 도서는 기존 수량을 더해서 업데이트합니다.

<br>

**4. 장바구니에 있는 도서 수량 업데이트 기능 추가**
- book id, 수량 정보를 받습니다.
- 기존 Cart 객체를 얻어 수량을 업데이트합니다.

<br>

**5. 장바구니에 있는 도서 삭제 기능 추가**
- user + book id에 해당하는 cart 데이터를 삭제합니다.

<br>

**6. 장바구니 조회 기능 추가**
- user id를 포함한 cart 데이터를 모두 얻어와 반환합니다.

<br>

**7. 장바구니 비우기 기능 추가**
- user id를 포함한 cart 데이터를 모두 삭제합니다.

<br>

**8. 테스트 코드 추가 (통과된 테스트는 체크 표시)**

Service/CartServiceTest

- [x] 장바구니 도서 추가 성공 - 장바구니에 존재하지 않는 도서
- [x] 장바구니 도서 추가 성공 - 장바구니에 이미 존재하는 도서
- [x] 장바구니 도서 추가 실패 - 존재하지 않는 도서
- [x] 장바구니 조회 성공
- [x] 장바구니 도서 수량 업데이트 성공
- [x] 장바구니 도서 수량 업데이트 실패 - 존재하지 않는 장바구니 상품
- [x] 장바구니 도서 삭제 성공
- [x] 장바구니 상품 전체 삭제 성공

<br>

테이블에 대한 질문이 있어서 erd 첨부합니다!

<img width="976" alt="스크린샷 2023-01-05 오전 4 32 27" src="https://user-images.githubusercontent.com/111635735/211625824-c10cc5e7-4b61-4ad2-90f0-a4e5345e60b0.png">

1. 테이블 cart를 cart_item으로 이름을 변경해도 괜찮을까요??
- 실제로 cart 테이블에 cart가 아닌 cart item이 저장이 돼서 테이블 이름도 cart_item으로 해줘야 되는 건지 고민이 됩니다..!

<br>

### **📌 TO-BE**
- 장바구니 선택 삭제 기능 추가

### 테스트
<!-- 본 변경사항이 테스트가 되었는지 기술해주세요 --> 
- [x] 테스트 코드
- [x] API 테스트

junga970 committed:
Feat: 장바구니 기능 추가

- #5

heyazoo1007 commented:
안녕하세요 팀 PR 2조에 김예진이라고 합니다 :) 코드가  깔끔하고 가독성이 높아 보기 편하네요. 많이 배우고 갑니다! 수고하셨고, 이번주도 화이팅입니다!

junga970 commented:
@heyazoo1007 좋게 봐주셔서 감사합니다! :grinning:

mingming-mentor commented on src/main/java/com/example/bookstore/service/CartService.java:
```diff
@@ -0,0 +1,91 @@
+		cartRepository.delete(cartItem);
+	}
+
+	public void DeleteCart(String userEmail) {
```
method명은 소문자로 시작하면 좋을 것 같아요

junga970 replied:
수정했습니다!

mingming-mentor commented on src/main/java/com/example/bookstore/service/CartService.java:
```diff
@@ -0,0 +1,91 @@
+
+	public void DeleteBookInCart(String userEmail, Long bookId) {
+
+		User user = userRepository.findByEmail(userEmail).get();
```
요 로직에서 아마 이미 앞쪽에서 authentication을 진행했기때문에 user가 없진 않을거라고 가정하시고 get()으로 바로 가지고오신거같아요. 요렇게 되면 로직이 인증과정에서 한번 유저체크를 하느라 db에 갔다오고, 그리고 이 과정에서 다시한번 유저를 가지고와야해서 2번의 user check가 일어나게 됩니다. 
가지고오신 user정보는 아래서 cart 조회시에 사용하는것으로 보이는데요
user 라는건 저희가 JPA객체로 cart를 표현할때 cart하위에 넣어준 값이고 실제 table에는 user_id만 column으로 있을거에요. 즉 조회시에 findByUserAndBookId로 하지 않고 findByUserIdAndBookId로 하셔도 무방합니다. 
(보니까 book은 bookId로 해서 넣어주셨네요ㅎㅎ )

junga970 replied:
@mingming-mentor 멘토님 질문이 있습니다!

유저체크1 -> `User user = userRepository.findByEmail(userEmail).get();`
유저체크2 -> `findByUserAndBookId(user, bookId);`

이렇게 2번 유저체크를 하는게 맞는건가용??
 `findByUserIdAndBookId(userId, bookId)` 처럼 
userId를 바로 넣어주면 유저체크를 할 일이 없는데,
User객체로 넣으면 id를 얻기 위해 유저체크를 한번 더 하게 된다는 말씀이신가용..??

mingming-mentor replied:
안녕하세요 @junga970 님! 요거 제가 jwt filter쪽 메소드를 착각하고 있었습니다. 지금 확인해보니 jwt token validation을 할때 token에 담긴 user 정보에 대해서 실제 db에 저장된 user정보인지는 따로 체크하고 있지 않네요. 
제가 생각했었던것은, jwt token validation시 db select를 통해 user가 실제 존재하는지 체크하는 로직 (1) 
그리고 `User user = userRepository.findByEmail(userEmail).get(); `를 통해서 또 유저를 가지고 오는 부분 (2)
해서 총 2번의 user select가 일어난다고 생각했습니다. 그래서 2번의 user 체크라고 말씀드렸어요. 
현재 코드를 보니 (1)의 경우가 존재하지 않아 1번만 발생하는 것으로 보이네요 :) 

user가 삭제되거나, 블락되거나 한 경우등을 위해서 jwt token validation시에 user select를 한번 해주기도 합니다. 이 부분은 필요에 따라 추가하셔도 좋을 것 같아요.

junga970 replied:
@mingming-mentor 
설명 감사합니다! 피드백 주신 대로, 토근 검증 시 유저 확인을 하도록 했습니다.
또한 컨트롤러 user.getUserName()에서 userEmail이 아닌 userId를 가져와 db조회 없이 바로 쓸 수 있게 수정했습니다!

mingming-mentor commented on src/main/java/com/example/bookstore/service/CartService.java:
```diff
@@ -0,0 +1,91 @@
+				.book(book)
+				.quantity(quantity)
+				.build();
+		}
```
```java
cartRepository.findByUserAndBookId(user, bookId).ifPresentOrElse(
    cartItem ->  {
        cartItem.setQuantity(cartItem.getQuantity() + quantity);
        cartRepository.save(cartItem);
    }, 
    () -> cartRepository.save(Cart.builder()
				.user(user)
				.book(book)
				.quantity(quantity)
				.build()); 
);
```
요렇게 ifPresentOrElse로 묶어서 짜셔도 좋을 것 같습니다.

junga970 replied:
수정했습니다! 덕분에 메서드 하나 알게 됐네요:grinning:

mingming-mentor reviewed (approved):
안녕하세요! 밍밍멘토입니다. 코드 잘보았습니다 :) 질문주신 사항에 대해서 

> 실제로 cart 테이블에 cart가 아닌 cart item이 저장이 돼서 테이블 이름도 cart_item으로 해줘야 되는 건지 고민이 됩니다..!

cart item이라는 이름이 더 직관적이라고 생각은 드네요. 바꾸시는 편이 더 좋을 것 같습니다. 수고 많으셨어요!

junga970 committed:
Refactor: Cart, jwt 관련 파일들과 UserDetailsServiceImpl 수정

1. CartController:
   1) userEmail -> userId 로 변경
   2) 메서드 이름 수정(오타)

2. Cart.java: CartItem 으로 이름 변경
- 데이터에 맞게 테이블 이름 변경

3. CartService: userEmail 관련 코드 삭제
- userId로 바로 사용

4. JwtTokenProvider: validateToken() 안에 user 확인 로직 추가
- service 에서 체크하지 않고 jwt 토큰 검증 시 체크

5. UserDetailsServiceImpl: 사용자 정보 반환 시 email -> userId(String) 으로 변경
- CartService 에서 email 을 통해 id를 얻는 로직을 제거해 db 조회를 최소화

6. 그 외:
- 위와 같은 이유들로 수정

- #5
